File: web/projects/portal/src/app/widget/tree/search-comparator.ts
/*
 * inetsoft-web - StyleBI is a business intelligence web application.
 * Copyright © 2024 InetSoft Technology (info@inetsoft.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import { TreeNodeModel } from "./tree-node-model";
import { SecurityTreeNode } from "../../../../../em/src/app/settings/security/security-tree-view/security-tree-node";
import { RepositoryTreeNode } from "../../../../../em/src/app/settings/content/repository/repository-tree-node";

export class SearchComparator {
   constructor(search: string) {
      this.searchStr = !!search ? search.toLowerCase() : search;
   }

   public searchSort(a: TreeNodeModel | SecurityTreeNode | RepositoryTreeNode,
                     b: TreeNodeModel | SecurityTreeNode | RepositoryTreeNode): number
   {
      let ad = this.getMatchDegree(a);
      let bd = this.getMatchDegree(b);

      return bd - ad;
   }

   public searchSortStr(a: string, b: string): number {
      let ad = this.getDegree(a);
      let bd = this.getDegree(b);
      return bd - ad;
   }

   private getMatchDegree(node: TreeNodeModel | SecurityTreeNode | RepositoryTreeNode) {
      var degree: number;
      if(node instanceof  SecurityTreeNode) {
          degree = this.getDegree(node.identityID.name.toLowerCase());
      }
      else {
          degree = this.getDegree(node.label.toLowerCase());
      }
      if(node.children == null) {
         return degree;
      }

      for(let child of node.children) {
         degree = Math.max(degree, this.getMatchDegree(child));
      }

      return degree;
   }

   private getDegree(name: string) {
      name = !!name ? name.toLowerCase() : name;

      if(name == this.searchStr) {
         return 5;
      }

      if(name.indexOf(" ") != -1) {
         let names = name.split(" ");

         for(let i = 0; i < names.length; i++) {
            if(names[i] == this.searchStr) {
               return name.startsWith(this.searchStr) && i == 0 ? 4 : 3;
            }
         }
      }

      if(name.startsWith(this.searchStr)) {
         return 2;
      }

      if(name.includes(this.searchStr)) {
         return 1;
      }

      return 0;
   }

   private searchStr: string = null;
}

File: core/src/main/java/inetsoft/web/portal/controller/SearchComparator.java
/*
 * inetsoft-core - StyleBI is a business intelligence web application.
 * Copyright © 2024 InetSoft Technology (info@inetsoft.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affrero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package inetsoft.web.portal.controller;

import inetsoft.util.Tool;
import inetsoft.web.portal.data.WorksheetBrowserInfo;
import inetsoft.web.portal.model.database.*;

import java.util.Comparator;
import java.util.List;

/**
 * SearchComparator.
 *
 * @since 13.4
 */
public class SearchComparator {
   private static class BaseComparator {
      public BaseComparator(String searchStr) {
         this.search = searchStr;
      }

      public int compare(String a, String b) {
         int ag = getMatchDegree(a);
         int bg = getMatchDegree(b);

         return bg - ag;
      }

      protected int getMatchDegree(String path) {
         String[] paths = path.split("/");
         int degree = 0;

         for(int i = 0; i < paths.length; i++) {
            String p = paths[i];
            degree = Math.max(degree, getDegree(p));
         }

         return degree;
      }

      /*
         The math degree for entry(search string is aa).
         4---the same(such as: aa)
         3---part has the same(such as: aa bb)
         2---start with(such as: aabb)
         1---has(such as: bbaa)
      */
      private int getDegree(String path) {
         String npath = path.toLowerCase();
         String searchStr = search.toLowerCase();

         if(Tool.equals(npath, searchStr)) {
            return 5;
         }

         if(npath.indexOf(" ") > 0) {
            String[] paths = npath.split(" ");

            for(int i = 0; i < paths.length; i++) {
               if(Tool.equals(paths[i].toLowerCase(), searchStr.toLowerCase())) {
                  return npath.startsWith(searchStr) && i == 0 ? 4 : 3;
               }
            }
         }

         int index = npath.indexOf(searchStr);

         if(index == 0) {
            return 2;
         }
         else if(index > 0) {
            return 1;
         }

         return 0;
      }

      private String search = null;
   }

   public static class StringComparator extends BaseComparator implements Comparator<String> {
      public StringComparator(String searchStr) {
         super(searchStr);
      }

      public int compare(String a, String b) {
         int ag = getMatchDegree(a);
         int bg = getMatchDegree(b);

         return bg - ag;
      }
   }

   public static class WorksheetBrowserComparator extends BaseComparator implements Comparator<WorksheetBrowserInfo> {
      public WorksheetBrowserComparator(String searchStr) {
         super(searchStr);
      }

      public int compare(WorksheetBrowserInfo a, WorksheetBrowserInfo b) {
         int ag = getMatchDegree(a.name());
         int bg = getMatchDegree(b.name());

         return bg - ag;
      }
   }

   public static class DataModelBrowserComparator extends BaseComparator implements Comparator<AssetItem> {
      public DataModelBrowserComparator(String searchStr) {
         super(searchStr);
      }

      public int compare(AssetItem a, AssetItem b) {
         int ag = getMatchItemDegree(a);
         int bg = getMatchItemDegree(b);

         return bg - ag;
      }

      /**
       * Get the max degree between item and children items.
       * @param a
       * @return
       */
      protected int getMatchItemDegree(AssetItem a) {
         int aggrete = getMatchDegree(a.getName());

         if(a instanceof LogicalModel) {
            LogicalModel logicalModel = (LogicalModel) a;
            List<LogicalModel> extendModels = logicalModel.getExtendModels();

            for(int i = 0; extendModels != null && i < extendModels.size(); i++) {
               if(extendModels.get(i) == null) {
                  continue;
               }

               aggrete = Math.max(aggrete, getMatchDegree(extendModels.get(i).getName()));
            }
         }

         if(a instanceof PhysicalModel) {
            PhysicalModel physicalModel = (PhysicalModel) a;
            List<PhysicalModel> extendViews = physicalModel.getExtendViews();

            for(int i = 0; extendViews != null && i < extendViews.size(); i++) {
               if(extendViews.get(i) == null) {
                  continue;
               }

               aggrete = Math.max(aggrete, getMatchDegree(extendViews.get(i).getName()));
            }
         }

         return aggrete;
      }
   }
}
